lib/ukdebug/gdbstub.c

```diff
@@ -1106,3 +1113,65 @@ int gdb_dbg_trap(int errnr, struct __regs *regs)
 
 	return r;
 }
+
+static int gdb_setup_backing_console(void)
+{
+	/* Ignore breakpoints if the stub is not enabled on the command line. */
+	if (gdb_backing_console_id == GDB_BACKING_CONSOLE_NONE)
+		return -ENXIO;
+
+	/* Ignore breakpoints if the backing console is not registered.
+	 * Maybe the backing console will be initialized later.
+	 */
+	if (gdb_backing_console_id >= uk_console_count()) {
+		uk_pr_warn("Console device %" __PRIu16 " not available\n",
+			   gdb_backing_console_id);
+		return -ENXIO;
+	}
+
+	gdb_backing_console = uk_console_get(gdb_backing_console_id);
+	UK_ASSERT(gdb_backing_console);
+
+	uk_console_init(&gdb_virt_console, "GDB virtual console",
+			&gdb_virt_ops, UK_CONSOLE_FLAG_STDOUT);
+	if (gdb_virt_console_on()) {
+		uk_console_register(&gdb_virt_console);
+	} else if (gdb_virt_console_auto()) {
+		/* Must check before diabling the flags (see below) */
+		if (gdb_backing_console->flags)
+			uk_console_register(&gdb_virt_console);
+	}
+
+	/* If we didn't disable the STDOUT and STDIN flags, calls to
+	 * `uk_console_[out|in]` that happen outside of the gdb stub
+	 * could mess up the gdb communications. E.g., a call to
+	 * `uk_console_out` could write some bytes to the device that
+	 * look like garbage to the gdb host on the other end.
+	 * We assume here that the backing device is registered with
+	 * `ukconsole` exactly  once. This means we can also assume that
+	 * disabling the STDOUT and STDIN flags makes sure that the
+	 * underlying device is never used except explicitly stated with
+	 * `uk_console_[out|in]_with(gdb_backing_console, ...)`.
+	 */
+	gdb_backing_console->flags &= ~UK_CONSOLE_FLAG_STDOUT;
+	gdb_backing_console->flags &= ~UK_CONSOLE_FLAG_STDIN;
+
+	return 0;
+}
+
+static int gdb_late_breakpoint(struct uk_init_ctx *ctx __unused)
+{
+	/* This is the first time that the stub is entered. */
+	UK_ASSERT(!gdb_backing_console);
```

Comment:
If we remove this assertion, we could call this init function multiple times right? I am thinking of patching it in a way that we try multiple times at various levels until we do not get `-ENXIO` anymore ... this way we would achieve kind of waiting for a console driver that is initializing late during init.

Reply:
It means that we need to skip successive calls when we connected once and keep retrying at each level if there was an issue before.

Reply:
> I am thinking of patching it in a way that we try multiple times at various levels until we do not get -ENXIO anymore ... this way we would achieve kind of waiting for a console driver that is initializing late during init.

In principle, this is exactly what we did before when we had the early/late breakpoints split. We have purposefully decided against this split, so I don't see any reason to re-introduce such a mechanism. The `gdb_late_breakpoint` (it should be renamed `gdb_init_breakpoint` or something similar) function runs, all console drivers have been initialized.

Reply:
What I meant is slightly different: During early init we had more difficulties with solving dependencies. The inittab is different because when it is called, the most basic system is initialized and should allow us without much adoption to call gdb stub at every init stage. The idea was that an earlier call allows earlier debugging (including device drivers) - especially if we use the early console with this. Because we may also depend on a serial console device that is initialized later, I suggested the retry to find the earliest place. This currently happens latest in `UK_INIT_CLASS_PLAT`, via the `lib/ukbus` triggered device scanning. You register at the end of the same class, so it should be fine.
However, I am thinking now that there might be an even more elegant way than multiple registrations to each initlevel, so probably worth to move such changes to a different PR.

Agree with the renaming (to `gdb_entry`). 👍🏻  `gdb_stub_entry` may work as well but I assume we will add more ways to enter the debugger later. So, it is good for now as is.

Reply:
To me that makes sense, as well as the proposal to look at it later. Perhaps in the scope of the topic of earlyboot, liblcpu and other items related to boot order.

Reply:
Agree.